test/hotspot/gtest/opto/test_xor_node.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "opto/addnode.hpp"
+#include "unittest.hpp"
+
+jint calc_max(const jint hi_0, const jint hi_1) {
+  return calc_xor_max<jint, juint>(hi_0, hi_1);
+}
+
+jlong calc_max(const jlong hi_0, const jlong hi_1) {
+  return calc_xor_max<jlong, julong>(hi_0, hi_1);
+}
+
+template <class S>
+void test_xor_bounds(S hi_0, S hi_1, S val_0, S val_1) {
+
+  // Skip out-of-bounds values for convenience
+  if(val_0> hi_0 || val_0 < S(0) || val_1 > hi_1 || val_1< S(0)) {
+    return;
+  }
+
+  S v = val_0 ^ val_1;
+  S max = calc_max(hi_0, hi_1);
+  EXPECT_LE(v, max);
+}
+
+template <class S>
+void test_exhaustive_values(S hi_0, S hi_1){
```

Comment:
You may want to test exhaustively for the `hi` values here, too. E.g:

    void test_exhaustive(S limit) {
        for (S hi0 = 0; hi0 <= limit; hi0++) {
            for (S hi1 = 0; hi1 <= limit; hi1++) {
                S max = calc_max(hi0, hi1);
                for (S v0 = 0; v0 <= hi0; v0++) {
                    for (S v1 = 0; v1 <= hi1; v1++) {
                        S v = v0 | v1;
                        EXPECT_LE(v, max);
                    }
                }
            }
        }
    }

Reply:
Was my intention that the test should have been doing that. I've cleaned it up to clarify it.